MoreLinq/Experimental/SelectAsync.cs

```diff
@@ -0,0 +1,548 @@
+
+                    var completedTask = await
+                        Task.WhenAny(tasks.Cast<Task>()
+                                          .Concat(cancellationTaskSource.Task));
+
+                    if (completedTask == cancellationTaskSource.Task)
+                    {
+                        // Cancellation during the wait means the enumeration
+                        // has been stopped by the user so the results of the
+                        // remaining tasks are no longer needed. Those tasks
+                        // should cancel as a result of sharing the same
+                        // cancellation token and provided that they passed it
+                        // on to any downstream asynchronous operations. Either
+                        // way, this loop is done so exit hard here.
+
+                        return;
+                    }
+
+                    var task = (Task<(T Input, TResult Result)>) completedTask;
+                    tasks.Remove(task);
+                    collection.Add(resultNoticeSelector(task.Result.Input, task.Result.Result));
+
+                    if (more && (more = e.MoveNext()))
+                        tasks.Add(taskSelector(e.Current).Select(r => (e.Current, r)));
+                }
+
+                collection.Add(endNotice);
+            }
+            catch (Exception ex)
+            {
+                cancellationTokenSource.Cancel();
+                collection.Add(errorNoticeSelector(ex));
+            }
+
+            collection.CompleteAdding();
+        }
+
+        static class SelectAsyncEnumerable
+        {
+            public static ISelectAsyncEnumerable<T>
+                Create<T>(
+                    Func<SelectAsyncOptions, IEnumerable<T>> impl,
+                    SelectAsyncOptions options = null) =>
+                new SelectAsyncEnumerable<T>(impl, options);
+        }
+
+        sealed class SelectAsyncEnumerable<T> : ISelectAsyncEnumerable<T>
+        {
+            readonly Func<SelectAsyncOptions, IEnumerable<T>> _impl;
+
+            public SelectAsyncEnumerable(Func<SelectAsyncOptions, IEnumerable<T>> impl,
+                SelectAsyncOptions options = null)
+            {
+                _impl = impl;
+                Options = options ?? SelectAsyncOptions.Default;
+            }
+
+            public SelectAsyncOptions Options { get; }
+
+            public ISelectAsyncEnumerable<T> WithOptions(SelectAsyncOptions options)
+            {
+                if (options == null) throw new ArgumentNullException(nameof(options));
+                return Options == options ? this : new SelectAsyncEnumerable<T>(_impl, options);
+            }
+
+            public IEnumerator<T> GetEnumerator() => _impl(Options).GetEnumerator();
+            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+        }
+
+        static class TupleComparer<T1, T2>
+        {
+            public static readonly IComparer<(T1, T2)> Key1 =
```

Comment:
I would name these `Item1` and `Item2` since `ValueTuple` does not have keys.

Reply:
This is designed for sorting, which is why I called them `Key1` and `Key2` but I guess if you found it confusing then perhaps best to keep parity. I'll rename.